rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+*forward* mapping of its sample times via the method `map_sample_times`, which accepts a function like the inverse of the one that 
+would be used in `Curve::reparametrize`.
+
+The associated resampling method is `resample_uneven`:
+```rust
+/// Resample this [`Curve`] to produce a new one that is defined by interpolation over samples
+/// taken at the given set of times. The given `sample_times` are expected to contain at least
+/// two valid times within the curve's domain range.
+///
+/// Irredundant sample times, non-finite sample times, and sample times outside of the domain
+/// are simply filtered out. With an insufficient quantity of data, a [`ResamplingError`] is
+/// returned.
+///
+/// The domain of the produced [`UnevenSampleCurve`] stretches between the first and last
+/// sample times of the iterator.
+fn resample_uneven(
+    &self,
+    sample_times: impl IntoIterator<Item = f32>,
+) -> Result<UnevenSampleCurve<T>, ResamplingError> { //... }
+```
+
+Note that there is no equivalent to `samples` for uneven spacing; however, these can be obtained easily using `Iterator::map`:
+```rust
+for (t, v) in (0..100).map(|x| x * x).map(|t| (t, my_curve.sample(t))) {
+    println!("Value {v} at time {t}.");
+}
+```
+The reason for this shortfall is that user expectations surrounding filtering, sorting, clamping, and so on are hard to predict. For example, 
+`sample` here could be replaced with `sample_clamped`, or the second `map` with `filter_map` and `sample` with `sample_checked`, among many other 
+variations. With time, perhaps a good default will become apparent and be added to the API.
+
+Finally, for some common math types whose interpolation is especially obvious and well-behaved (and enshrined in a trait), there are
+convenience methods `resample_auto` and `resample_uneven_auto` that use this type-inferred interpolation automatically. The expectation is not
+that interpolation in user-space will commonly use this trait and the associated methods, since the bar for having a valid trait implementation
+is rather high; rather, in domains where weaker interpolation notions are prevalent (e.g. animation), the expectation is that consumers will
+primarily go through `SampleCurve` or `UnevenSampleCurve` or define their own curve constructions entirely.
+
+### Combining curves
+
+There are a couple of common ways of combining curves that are supported by the Curve API. The first of these is `compose`, which appends two curves together end-to-end:
+```rust
+/// Create a new [`Curve`] by composing this curve end-to-end with another, producing another curve
+/// with outputs of the same type. The domain of the other curve is translated so that its start
+/// coincides with where this curve ends. A [`CompositionError`] is returned if this curve's domain
+/// doesn't have a finite right endpoint or if `other`'s domain doesn't have a finite left endpoint.
+fn compose<C>(self, other: C) -> Result<impl Curve<T>, CompositionError> { //... }
```

Comment:
> I feel like this function needs a bit more thought. What is the behavior if `self.end() != other.begin()`?

Right now, it just translates the second curve so that its start coincides with the end of the first. If you mean in terms of values — there are no guarantees: a `Curve` isn't necessarily continuous, and in general this operator will combine two continuous curves into a discontinuous one.

> What is the resulting type of composing curves that themselves are composed of curves?
> 
> > My suggestion: have that be `PiecewiseCurve<impl Curve<T>>`. This allows composing a variable number of curves at runtime.

Yeah, there is probably more to be done here; the goal of the minimal version of end-to-end composition in the API is to work with arbitrary curves, so the output type contains those of the input curves as generic parameters. The story for longer sequences of curves is a little thornier, since we don't really have `HList`s. One idea is to do what you're saying with piecewise constructions, but this only works if every curve in the sequence has exactly the same type. In its most general form, there probably isn't any way to get around using `Box<dyn Curve<T>>` as the curve type in such a thing presently.

Reply:
For the generic case dynamic dispatch is practically unavoidable. Even a Composed<Curve<A>,Curve<B>> has a function call stuck behind a branch, that the compiler probably can't do much with.

I'm not sure whether the PiecewiseCurve I suggest here can actually be implemented in rust. If I'm correct, you cannot have:

`fn chain(self, other: Curve<T>) -> PiecewiseCurve<Curve<T>;` in the generic case; and
`fn chain(self, other: PiecewiseCurve<Curve<T>>) -> PiecewiseCurve<Curve<T>;` as specialization. 

This may be possible with either negative trait bounds or specialization, but both are still unstable features and might still not provide the needed flexibility.

Reply:
You can do something like that with the caveat that you use an opaque `impl Curve<T>` return type in the trait definition (and do the specialization on `self` rather than `other`). I have personally gone back and forth with these sorts of optimizations, since they basically get destroyed the moment you do anything other than call the same method twice.